src/analysis/ring_seminorm.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import analysis.seminorm
+
+/-!
+# Seminorms and norms on rings
+
+This file defines seminorms and norms on rings. These definitions are useful when one needs to
+consider multiple (semi)norms on a given ring.
+
+## Main declarations
+
+For a ring `R`:
+* `ring_seminorm`: A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes
+  nonnegative values, is subadditive and submultiplicative and such that `f (-x) = f x` for all
+  `x ∈ R`.
+* `ring_norm`: A seminorm `f` is a norm if `f x = 0` if and only if `x = 0`.
+
+## References
+
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+ring_seminorm, ring_norm
+-/
+
+noncomputable theory
+
+set_option old_structure_cmd true
+
+open_locale nnreal
+
+variables {R S : Type*} (x y : R) (r : ℝ)
+
+/-- A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes nonnegative
+  values, is subadditive and submultiplicative and such that `f (-x) = f x` for all `x ∈ R`. -/
+structure ring_seminorm (R : Type*) [non_unital_ring R]
+  extends add_group_seminorm R :=
+(mul_le' : ∀ x y : R, to_fun (x * y) ≤ to_fun x * to_fun y)
+
+attribute [nolint doc_blame] ring_seminorm.to_add_group_seminorm
+
+namespace ring_seminorm
+
+section non_unital_ring
+
+variables [non_unital_ring R]
+
+instance zero_hom_class : zero_hom_class (ring_seminorm R) R ℝ :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_zero := λ f, f.map_zero' }
```

Comment:
As future work: it would be good to define an `add_group_seminorm_class` too, so that `ring_seminorm` can provide an instance of it, and then you can use them slightly more interchangeably

Reply:
I'll add this as a TODO.